jvm/src/main/scala/org/portablescala/reflect/Reflect.scala

```diff
@@ -0,0 +1,163 @@
+   *  Moreover, the class must not be abstract, nor be a local class (i.e., a
+   *  class defined inside a `def` or inside an anonymous function). Inner
+   *  classes (defined inside another class) are supported.
+   *
+   *  If the class cannot be found, either because it does not exist,
+   *  was not `@EnableReflectiveInstantiation` or was abstract or local, this
+   *  method returns `None`.
+   *
+   *  This method is equivalent to calling
+   *  {{{
+   *  Reflect.lookupInstantiatableClass(fqcn, Reflect.getClass.getClassLoader)
+   *  }}}
+   *
+   *  @param fqcn
+   *    Fully-qualified name of the class
+   */
+  def lookupInstantiatableClass(fqcn: String): Option[InstantiatableClass] =
+    lookupInstantiatableClass(fqcn, this.getClass.getClassLoader)
+
+  /** Reflectively looks up an instantiatable class.
+   *
+   *  The class or one of its super types (classes or traits) must be annotated
+   *  with
+   *  [[org.portablescala.reflect.annotation.EnableReflectiveInstantiation @EnableReflectiveInstantiation]].
+   *  Moreover, the class must not be abstract, nor be a local class (i.e., a
+   *  class defined inside a `def` or inside an anonymous function). Inner
+   *  classes (defined inside another class) are supported.
+   *
+   *  If the class cannot be found, either because it does not exist,
+   *  was not `@EnableReflectiveInstantiation` or was abstract or local, this
+   *  method returns `None`.
+   *
+   *  @param fqcn
+   *    Fully-qualified name of the class
+   *
+   *  @param loader
+   *    Class loader to use to load the class
+   */
+  def lookupInstantiatableClass(fqcn: String,
+      loader: ClassLoader): Option[InstantiatableClass] = {
+    load(fqcn, loader).filter(isInstantiatableClass).map(new InstantiatableClass(_))
+  }
+
+  private def isModuleClass(clazz: Class[_]): Boolean = {
+    try {
+      val fld = clazz.getField("MODULE$")
+      clazz.getName.endsWith("$") && (fld.getModifiers & Modifier.STATIC) != 0
+    } catch {
+      case _: NoSuchFieldException => false
+    }
+  }
+
+  private def isInstantiatableClass(clazz: Class[_]): Boolean = {
+    /* A local class will have a non-null *enclosing* class, but a null
+     * *declaring* class. For a top-level class, both are null, and for an
+     * inner class (non-local), both are the same non-null class.
+     */
+    def isLocalClass: Boolean =
+      clazz.getEnclosingClass() != clazz.getDeclaringClass()
+
+    (clazz.getModifiers() & Modifier.ABSTRACT) == 0 &&
+    clazz.getConstructors().length > 0 &&
+    !isModuleClass(clazz) &&
+    !isLocalClass
+  }
+
+  private def load(fqcn: String, loader: ClassLoader): Option[Class[_]] = {
+    try {
+      /* initialize = false, so that the constructor of a module class is not
+       * executed right away. It will only be executed when we call
+       * `loadModule`.
+       */
+      val clazz = Class.forName(fqcn, false, loader)
+      if (inheritsAnnotation(clazz)) Some(clazz)
```

Comment:
Do we also need to check that the class is public? Note the documentation of `Class.forName`:

> Note that this method does not check whether the requested class is accessible to its caller.

Reply:
Apparently not. Both JVM and JS agree that a private class is discoverable. See the additional commit with a test case.

Reply:
:O is that really what we want it to do?

Reply:
There's nothing else we *can* do because in the JVM bytecode such a class is public anyways.

Reply:
I see... So Scala never emits a non-public class?

Reply:
Apparently not ...